diff --git a/internal/streams/add_consumer.go b/internal/streams/add_consumer.go
index 7400ce6..1f5c854 100644
--- a/internal/streams/add_consumer.go
+++ b/internal/streams/add_consumer.go
@@ -11,7 +11,11 @@ func (s *Stream) AddConsumer(cons core.Consumer) (err error) {
 	// support for multiple simultaneous pending from different consumers
 	consN := s.pending.Add(1) - 1
 
-	var prodErrors = make([]error, len(s.producers))
+	s.mu.Lock()
+	producers := append([]*Producer(nil), s.producers...)
+	s.mu.Unlock()
+
+	var prodErrors = make([]error, len(producers))
 	var prodMedias []*core.Media
 	var prodStarts []*Producer
 
@@ -21,7 +25,7 @@ func (s *Stream) AddConsumer(cons core.Consumer) (err error) {
 		log.Trace().Msgf("[streams] check cons=%d media=%s", consN, consMedia)
 
 	producers:
-		for prodN, prod := range s.producers {
+		for prodN, prod := range producers {
 			// check for loop request, ex. `camera1: ffmpeg:camera1`
 			if info, ok := cons.(core.Info); ok && prod.url == info.GetSource() {
 				log.Trace().Msgf("[streams] skip cons=%d prod=%d", consN, prodN)
